### PR TITLE
WT-6673 RTS Fix inconsistent checkpoint

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
Rollback to stable to fix the inconsistent checkpoint by removing updates outside of the checkpoint snapshot.

- All the RTS mechanisms will be modified for an additional transaction id check(only during recovery) in addition to the current timestamp check on each page/record level.
- During RTS in recovery reset transaction ids based on write gen, re-initialize the connection's base write generation after rollback to stable, and update write_gens and base_write_gen of all the open dhandles to the connection's base write generation.
- Add a new inconsistent checkpoint stats counter.
- Add a new test to simulate the inconsistent checkpoint behavior.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
@tetsuo-cpp The way we reset the transaction ids is the logic that is modified from the last committed version as the older approach of not resetting transaction ids during recovery led to reading stale transaction ids.

File: src/txn/txn_recover.c [hunk 1]
```diff
@@ -781,6 +783,7 @@ __wt_txn_recover(WT_SESSION_IMPL *session, const char *cfg[])
          * earlier time.
          */
         WT_ERR(__recovery_file_scan(&r));
+
```
Comment (kommiharibabu) [inline src/txn/txn_recover.c:786]: remove this extra line.

File: src/txn/txn_recover.c [hunk 2]
```diff
@@ -796,6 +799,9 @@ __wt_txn_recover(WT_SESSION_IMPL *session, const char *cfg[])
         goto done;
     }
 
+    __wt_verbose(
```
Comment (kommiharibabu) [inline src/txn/txn_recover.c:802]: Remove these verbose messages.

File: src/txn/txn_recover.c [hunk 3]
```diff
@@ -806,6 +812,8 @@ __wt_txn_recover(WT_SESSION_IMPL *session, const char *cfg[])
      * recovery can continue. Other errors remain errors.
      */
     if (!was_backup) {
+        __wt_verbose(
```
Comment (kommiharibabu) [inline src/txn/txn_recover.c:815]: Here too.

File: src/btree/bt_handle.c [hunk 4]
```diff
@@ -524,9 +524,24 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
...
     /* If this is the first time opening the tree this run. */
     if (F_ISSET(session, WT_SESSION_IMPORT) || ckpt->run_write_gen < conn->base_write_gen)
-        btree->base_write_gen = btree->run_write_gen = btree->write_gen;
+        btree->run_write_gen = btree->write_gen;
     else
-        btree->base_write_gen = btree->run_write_gen = ckpt->run_write_gen;
+        btree->run_write_gen = ckpt->run_write_gen;
+
+    /*
+     * Don't reset the transaction ids in recovery because rollback to stable in addition to stable
```
Comment (kommiharibabu) [inline src/btree/bt_handle.c:532]: Can we update the comment why we skip logged tables and metadata as RTS doesn't happen on them?

**Review by kommiharibabu — APPROVED** [review]
LGTM. A couple of verbose or comment fixes

**Comment by raviprakashgiri29** [comment]
@tetsuo-cpp Can you please review these changes?

**Comment by tetsuo-cpp** [comment]
@raviprakashgiri29 
Can you please write some information about how this is different from the last PR? All I saw is that the last one got reverted and now it's back again.

**Comment by raviprakashgiri29** [comment]
> @raviprakashgiri29
> Can you please write some information about how this is different from the last PR? All I saw is that the last one got reverted and now it's back again.

@tetsuo-cpp This ticket was reverted because in the old changes transaction ids were never reset during RTS and due to which RTS ends up removing some stale data. 
Now with the new change [here](https://github.com/wiredtiger/wiredtiger/pull/6303/files#diff-8db04637c9fae0129b8e6c9e480b19da95b1196f43f0fb7c101d381715525590R536), transaction ids get reset based on the last checkpointed run write generation, and transaction ids are retained only on the pages that are written after the restart.

File: src/btree/bt_handle.c [hunk 5]
```diff
@@ -524,9 +524,26 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
...
+     * timestamp also depends on transaction ids from the page that are read into cache to decide if
+     * an update needs to be rolled back.
+     *
+     * To achieve that, use the last checkpointed run write generation number as base write
```
Comment (tetsuo-cpp) [inline src/btree/bt_handle.c:536]: I don't get the comment. Above it says that you don't want to wipe transaction ids. And here it's saying that we're only retaining transaction ids for pages written after the restart.

File: src/btree/bt_handle.c [hunk 6]
```diff
@@ -524,9 +524,26 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
...
+     * generation number. The transaction ids are retained only on the pages that are written after
+     * the restart.
+     *
+     * Rollback to stable does not operate on logged tables and metadata, so it is skipped.
+     */
+    if (!F_ISSET(conn, WT_CONN_RECOVERING) || WT_IS_METADATA(btree->dhandle) ||
+      __wt_btree_immediately_durable(session))
+        btree->base_write_gen = btree->run_write_gen;
+    else
+        btree->base_write_gen = ckpt->run_write_gen;
```
Comment (tetsuo-cpp) [inline src/btree/bt_handle.c:546]: Is the idea that you want the `btree->base_write_gen` to be temporarily lower than any of the cells we're reading in via rollback to stable? Is this effectively the same as just setting it to 0?

Comment (kommiharibabu) [reply src/btree/bt_handle.c:546]: No. The earlier problem of RTS was reading the pages that are modified before the last restart.
Eg steps:
1. A page was written with txn id 100
2. restart -1
3. some more operations but the original page was not touched.
4. restart -2 --> This restart is reading the page with txnid 100 and remove its updates as part of RTS.

if we use 0, we will get the same problem, so we want to set the base write generation number as the run write generation number at restart -1, so any page that is modified after restart-1 will have transaction ids and rest of the pages will get reset.

Comment (tetsuo-cpp) [reply src/btree/bt_handle.c:546]: That's interesting. I think I get it.

File: src/btree/bt_handle.c (hunk 5 above)
Comment (raviprakashgiri29) [reply src/btree/bt_handle.c:536]: I will fix the comment now. This was actually from the last changes.

File: src/txn/txn_recover.c (hunk 1 above)
Comment (raviprakashgiri29) [reply src/txn/txn_recover.c]: Fixed

File: src/txn/txn_recover.c (hunk 3 above)
Comment (raviprakashgiri29) [reply src/txn/txn_recover.c]: Fixed

File: src/txn/txn_recover.c (hunk 2 above)
Comment (raviprakashgiri29) [reply src/txn/txn_recover.c]: Fixed

File: src/btree/bt_handle.c (hunk 4 above)
Comment (raviprakashgiri29) [reply src/btree/bt_handle.c]: Updated the comments to describe with reason to skip metadata and logged tables.

**Review by tetsuo-cpp — APPROVED** [review]

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**